### ISSUE TITLE
Proposed API for Bluesky service (for discussion)

### ISSUE THREAD

**Issue by dmgav** [bluesky/bluesky#1561]
Related issue: https://github.com/bluesky/bluesky/issues/1555

Communication with the server will be performed using two (REQ-REP and PUB-SUB) sockets.

### `status` API

Current status of the server. The status may be requested even when remote control is not enabled. The status is also published on the PUB-SUB socket periodically and each time **remote_enabled** and **process_idle** is changed.  

**Returns:**

- **remote_enabled**: boolean, indicates if the remote applications are allowed to control the Run Engine. If remote control is disabled, the remote applications are still able to request status of the server, document stream URI and the list of runs for the currently running plan. 

- **re_state**: current state of Run Engine, obtained directly from the Run Engine.

- **process_idle**: boolean, True if the server is ready to accept the requests, such as a request to start a new plan or to **resume**/**stop**/**abort**/**halt** a paused plan. Check **re_state** state for additional information. 

### `start_plan` API
 
Work similarly to '[queue_item_execute](https://blueskyproject.io/bluesky-queueserver/re_manager_api.html#method-queue-item-execute)' API of Queue Server.

**Parameters:**

- **item**: dictionary representation of a plan (see below).

**Returns:**

- **success**: boolean, True if the request was accepted, False otherwise (e. g. another plan is currently running). Success does not mean that the plan is successfully started. 

- **msg**: string, error message if the request failed. 

### `pause_plan` API

Initiate pause of the currently running plan. 

**Parameters:**

- **immediate**: boolean, optional, default False. Set True to initiate an immediate pause of the Run Engine. 

**Returns:**

- **success**: boolean. *True* if the request was processed correctly, *False* otherwise. 

- **msg**: string, error message if the request failed. 

 
### `resume_plan`, `stop_plan`, `abort_plan`, `halt_plan` API

Resume, stop, abort and halt a paused plan. 

**Returns:**

- **success**: boolean. *True* if the request was processed correctly, *False* otherwise. 

- **msg**: string, error message if the request failed. 

 
 
### `get_run_list` API

Returns the list of runs produced by currently running plan maintained by Run Engine. The list could be loaded at any time, even if remote control is not enabled.  

The full list of runs can be loaded at any time, which allows for applications that are started in the middle of the run to start properly. The list is also published on PUB-SUB socket each time there is a change and at the time the plan is completed (successfully or not). The list is erased once plan execution is completed. The run list persists through pause/resume sequence. 

The API is a simplified version of ‘[re_runs](https://blueskyproject.io/bluesky-queueserver/re_manager_api.html#re-runs)’ API of Queue Server. 

**Returns:**

- **success**: boolean. True if the request was processed correctly, False otherwise. 

- **msg**: string, error message if the request failed. 

- **runs**: *list(dict)*, a list of dictionaries, each dictionary contains information of a run generated by currently executed plan. Most plans produce a single run, so the list will contain only one dictionary. Format of the list element:  
  
  ``` 
  {"uid": uid, "is_open": True, "exit_status": None},  
  ```
  
  where `uid` is a run UID, `is_open` is *True* if run is open (start document was emitted) and *False* if the run is closed (stop 
  document was emitted), `exit_status` is a string representation of the exit status from the stop document, otherwise *None*. 

### `get_document_stream_uri` API

Returns the URI used for publishing Bluesky document stream. The stream name should be used by clients to monitor documents, e. g. for live plotting. The location can be requested even if remote operation is not enabled. 

**Returns:**

- **uri**: string, such as `'zqm://x.y.v:port'` or `'kafka://some_cluster.v:topic'`. Returns `''` (empty string) or *None* if the server does not know the stream URI. This would typically indicate an issue with server configuration.

## PUB-SUB 0MQ socket

The following messages are published (suggested topic name: `'bluesky_monitoring'`): 

1. **Status** of the server (information returned by REQ-REP **status** API). Status is published periodically (e.g. once per second, it could be parameterized) and each time **remote_enabled** and **process_idle** fields are changed. The status may be requested using REQ-REP **status** API at any time. There is no clear way to immediately publish the status with each change of **re_state**. 

2. **Reports** are published each time the plan is completed (successfully or not) or paused. The report is not stored, so there is no REQ-REP API to load it. The report contains the following fields:  

    - **plan_state**: `'completed'`, `'paused'`, `'stopped'`, `'aborted'`, `'halted'`, `'failed'` (exception was raised while the plan is being started or Run Engine crashed) 

    - **result**: value returned by the plan if it exits normally, `''` (empty string) or *None* if an exception was raised. 

    - **traceback**: traceback if an exception was raised, otherwise `''` (empty string).

3. **Run list**. The full list of runs is published whenever a new run is started or finished and at the completion of a plan.  
 

## Representation of Plans 
 
Each submitted plan is represented as a dictionary, e. g.: 

```
{"name": "count", "args": [["det1", "det2"]], "kwargs": {"num": 5, "delay": 1}} 
```
 
- `'name'` - plan name represented as a string.

- `'args'` - a list of plan args. 

- `'kwargs'` - a dictionary of plan kwargs.

- `'meta'` - a dictionary of plan metadata, which is passed to RE call (as `**meta`). 

The dictionary may contain extra unused parameters, which are ignored (e.g. `'item_type'`, `'user'`, `'user_group'` used by Queue Server). Only `'name'`, `'args'`, `'kwargs'` and `'meta'` are used.  

 
## Representation of Plan and Device Objects Passed as Plan Parameters.

Plans and devices are always represented as strings (e. g. `'count'`, `'det1'`, `['det1', 'det2']`, `'det1.value'` etc.) The client application has no access to the execution workspace and cannot operate with plan or device objects directly. The plan and device names are converted to plan and device objects by the worker before they are passed to the plan. 

Queue Server operates with lists of existing/allowed plans and devices, which are not going to be available in this worker. Therefore, Queue Server code cannot be used directly. The following approach is compatible with the default handling of parameters in Queue Server: 

- Before starting a plan, the worker binds values to parameters using plan signature.  

- The worker checks if string values of parameters without type annotation, are names of Python objects in execution namespace by attempting to convert string values of parameters without type annotation to an object from the namespace (by calling 'eval' function).  

- If conversion fails (no such object), then the value is passed as a string.  

- If a value is a list (iterable, but not a string) or a dictionary (mapping), the worker attempts to convert each list element or a values (not keys) of all dictionary element to objects from namespace.  

- For more complicated objects (e. g. list of lists of dictionaries etc.), the conversion is performed iteratively until the level of base types is reached.  

- In some case, a name of an existing Python object need to be passed to a plan as a string (without conversion). The conversion may be disabled by specifying type annotation for the parameter: 

  ``` 
  def plan1(detectors): 
      # Default behavior, detector names are converted to reference to Python objects 
      ... 
 
  def plan2(detector_names: list): 
      # Type annotation disables conversion. Values are passed as string 
      ... 
 
  def plan2(detector_names: typing.Any): 
      # Type annotation disables conversion. Values are passed as string 
      ... 
  ```
 
### Notes: 

The proposed conversion algorithm allows passing names of variables existing in the execution namespace. Those names are then converted to reference to the respective objects and passed to plans. This is currently not supported by the Queue Server, but could be useful, especially in interactive workflows. 

The values for some parameters that require callables (e. g. **per_step** parameter) could be passed as strings and then converted to references to callable objects.  

Execution of arbitrary code could be prevented by first validating the string with regular expression and ignoring all the strings that are not valid Python variables or Python variables separated by dots (access to object attributes).

**Comment by danielballan** [comment]
Thanks @dmgav. I'll reserve detailed comments for after others get a chance to weigh in.

I'll inject a comment on what we are broadly hoping to achieve with this discussion. Our goal is to build, together, the core of a "service" layer into the bluesky repo itself, so that we can:

1. Standardize on a sans I/O JSON API for submitting and monitoring Bluesky plans that can be shared across various larger, full-featured developments at NSLS2, Diamond, PSI, etc.
2. Provide a basic service included in bluesky that can be used to do out-of-process control of the RunEngine. This may be used by simple applications (e.g. simple GUIs prototyped by scientist post-docs) or used as a stepping stone while building up toward larger ones developed by facilities.

**Comment by stan-dot** [comment]
@danielballan arguably the services wrapping runengine haven't been built in this repo - such as https://github.com/bluesky/bluesky-queueserver 
and https://github.com/DiamondLightSource/blueapi

and they seem to work alright without a standardized JSON API.

should we close this issue?

**Comment by stan-dot** [comment]
note: if this is closed for such reasons, the related issue https://github.com/bluesky/bluesky/issues/1555 
should be closed too
